### ISSUE TITLE
squashed migration broke django-reversion upgrade.

### ISSUE THREAD

**Issue by xuchen81** [etianen/django-reversion#559]
Issue:
Django version: 1.9.7
django-reversion version: 2.0.5
I was using django-reversion 1.8.0, and there is no "db" columns in reversion_version table.
Now I am upgrading to version 2.0.5, it adds the column "db" in migration 3, but since there is also a squashed migration file there, when I run python manage.py migrate --fake-initial, it will fake the squashed migration and ignore all the other migrations in the folder, and the migration 2,3,4 will never be ran, so "db" column will not be added to the existing reversion_version table.

Suggested solution:
Remove squashed migration.

Thanks.

**Comment by etianen** [comment]
Why are you faking the initial migration?

This isn't an issue with django-reversion. It *might* be an issue with Django's migration framework.

**Comment by alukach** [comment]
We are also getting this error.  I believe that the issue is based on the fact that in `reversion/migrations/0003_auto_20160601_1600.py`, the code is both a) adding the `db` column (lines [94-98](https://github.com/etianen/django-reversion/blob/master/reversion/migrations/0003_auto_20160601_1600.py#L94-L98)); and then b) using the `db` column (line [66](https://github.com/etianen/django-reversion/blob/master/reversion/migrations/0003_auto_20160601_1600.py#L66)).

From the [RunPython section of the Django Migration docs](https://docs.djangoproject.com/en/1.9/ref/migration-operations/#django.db.migrations.operations.RunPython), it states:

> On databases that do support DDL transactions (SQLite and PostgreSQL), **RunPython** operations do not have any transactions automatically added besides the transactions created for each migration (the **atomic** parameter has no effect on these databases). Thus, on PostgreSQL, for example, you should avoid combining schema changes and **RunPython** operations in the same migration or you may hit errors like **OperationalError: cannot ALTER TABLE "mytable" because it has pending trigger events**.

Basically, you can't create a column and then use it in the same migration with a database that supports DDL.  My guess is that @xuchen81 is using either SQLite or PostgreSQL and @etianen is using MySQL or Oracle.

I'm unsure of where we should go from here.  Possibly break out the `RunPython` code into a separate migration named `0003_set_version_db` and add that as a dependency for `0004_auto_20160611_1202.py`?

**Comment by etianen** [comment]
This isn't to do with combining a RunPython with a schema migration. I run PostgreSQL on all my production sites. If that was the issue, you'd get the specific error "OperationalError: cannot ALTER TABLE "mytable" because it has pending trigger events".

@alukach - are you also faking the initial migration?

**Comment by xuchen81** [comment]
I need to fake the migration, because I am upgrading django from 1.6 to 1.9 too, and back when it was at 1.6, there is no django native migration feature, so there is no table django_migrations.

This is what I needed to do, remove all the south migration files, upgrade django from 1.6 to 1.9.7, and then python manage.py makemigrations to create all the initial migrations to keep up with my existing tables. At this point, I needed to run python manage.py migrate --fake-initial to create django_migrations table, and put those initial migrations in the table. That's why.

**Comment by alukach** [comment]
@etianen We are not faking migrations.  Will see if I can reproduce the issue.  My suggestion about breaking out the `RunPython` code into a separate migration did not work, not sure why...

**Comment by etianen** [comment]
Try disabling the set_version_db migration, then see if migration 0003 completes. If it does, see if the db column actually exists on the Version table.

Something odd is happening here. These migrations worked to upgrade my own Version database tables, so I'm not sure what's up.

**Comment by etianen** [comment]
If I take out the squashed migration, I can still run all the migrations just fine on all supported combinations of Django and Python, and the db column exists at the point the RunPython migration runs.

Maybe this is some postgres bug? I'm running 9.5 here.

**Comment by xuchen81** [comment]
I am using mysql.

**Comment by xuchen81** [comment]
To reproduce the bug, you should have django 1.6 database structure, with django-reversion 1.8.0 installed, make sure you dont have django_migrations table, and your reversion_version table looks like:
`mysql> show columns from reversion_version;
+-----------------+--------------+------+-----+---------+----------------+
| Field           | Type         | Null | Key | Default | Extra          |
+-----------------+--------------+------+-----+---------+----------------+
| id              | int(11)      | NO   | PRI | NULL    | auto_increment |
| revision_id     | int(11)      | NO   | MUL | NULL    |                |
| object_id       | longtext     | NO   |     | NULL    |                |
| object_id_int   | int(11)      | YES  | MUL | NULL    |                |
| content_type_id | int(11)      | NO   | MUL | NULL    |                |
| format          | varchar(255) | NO   |     | NULL    |                |
| serialized_data | longtext     | NO   |     | NULL    |                |
| object_repr     | longtext     | NO   |     | NULL    |                |
+-----------------+--------------+------+-----+---------+----------------+
8 rows in set (0.00 sec)`

and then install django-reversion 2.0.5, and django version 1.9.7
run:
python manage.py migrate --fake-initial,

now, go back to your database, you will see no change on reversion_version table.

**Comment by gilmrjc** [comment]
Ok I think the problem is you are faking "all" migrations, but you don't have them. I suppose you have migration 0001 and 0002. Run manage.py migrate --fake reversion 0002 to get into that stage, and then just do manage.py migrate reversion to get the new migrations applied. After that run with the fake-inital flag to fake all the other aps.

**Comment by etianen** [comment]
I agree. That sounds like it would work.

**Comment by alukach** [comment]
@xuchen81 Any chance that you're using a multiple-db setup on Django? I think we've solved our problem: we have our data fragmented between multiple databases. We run migrations on each database with the `--database` flag.  The migration in `0002` doesn't use the `.using` function, updating the code to `Version.objects.using(db_alias).filter(db_query).update(db=db)` helped us.

**Comment by etianen** [comment]
@xuchen81 - is this a general fix that you could share with a pull request?

**Comment by alukach** [comment]
@etianen See the above PR.

@xuchen81 any chance that you could give that a try?  You can install this fork via `pip install git+git://github.com/alukach/django-reversion.git@master`

**Comment by etianen** [comment]
This is now merged into master.

**Comment by elpaquete** [comment]
We are having the same issue that @xuchen81 is having. From our understanding the solution proposed also should work, but from what I could read in other issues it seems to fix a MySQL error.

In our case, we had a DB in Postgres with Django 1.6. To upgrade we need to fake the initial migrations in all apps. With the original 0001_initial this works perfectly, since the schema is the same as the latest available in the south migrations.

Now, with the squashed migrations this does not work. Changes in other migrations (particularly 0003) are not executed, resulting in an inconsistent state between the models and the DB.

```
dbname=# select * from django_migrations where app='reversion';
 id |    app    |                 name                  |            applied            
----+-----------+---------------------------------------+-------------------------------
 21 | reversion | 0001_initial                          | 2016-07-25 20:36:42.276982+00
 22 | reversion | 0002_auto_20141216_1509               | 2016-07-25 20:36:42.450425+00
 23 | reversion | 0003_auto_20160601_1600               | 2016-07-25 20:36:42.51696+00
 24 | reversion | 0004_auto_20160611_1202               | 2016-07-25 20:36:42.535391+00
 29 | reversion | 0001_squashed_0004_auto_20160611_1202 | 2016-07-25 20:36:43.135753+00
(5 rows)

dbname=# select * from south_migrationhistory where app_name='reversion';
 id | app_name  |                 migration                  |            applied            
----+-----------+--------------------------------------------+-------------------------------
 20 | reversion | 0001_initial                               | 2014-08-14 22:51:09.220035+00
 21 | reversion | 0002_auto__add_field_version_type          | 2014-08-14 22:51:09.326007+00
 22 | reversion | 0003_auto__add_field_version_object_id_int | 2014-08-14 22:51:09.37565+00
 23 | reversion | 0004_populate_object_id_int                | 2014-08-14 22:51:09.40069+00
 24 | reversion | 0005_auto__add_field_revision_manager_slug | 2014-08-14 22:51:09.489379+00
 25 | reversion | 0006_remove_delete_revisions               | 2014-08-14 22:51:09.508733+00
 26 | reversion | 0007_auto__del_field_version_type          | 2014-08-14 22:51:09.519611+00
 38 | reversion | 0008_auto__add_index_revision_date_created | 2016-07-12 17:49:24.73843+00
(8 rows)

dbname=# \d reversion_revision
                                     Table "public.reversion_revision"
    Column    |           Type           |                            Modifiers                            
--------------+--------------------------+-----------------------------------------------------------------
 id           | integer                  | not null default nextval('reversion_revision_id_seq'::regclass)
 date_created | timestamp with time zone | not null
 user_id      | integer                  | 
 comment      | text                     | not null
 manager_slug | character varying(200)   | not null

```

In this case, the manager_slug field is still present, resulting in `null value in column "manager_slug" violates not-null constraint` errors.

The other issue is that the squashed migration was released simultaneously with the other migrations. So the initial migration (squashed) is actually different from the south one in 2.x releases.

A solution could be to install 1.10.2, migrate (fake-initial), then upgrade to 2.0.6 and migrate again. We are about to try that, I'll post the results.

**Comment by elpaquete** [comment]
Using 1.10.2 to bump the Django version works well.

If updating Django and reversion, the following works (although it needs some small adjustments in the code for each step).

```
# If you didn't have the latest version that does not require Django 1.7
$ pip install django-reversion==1.8.7
$ python manage.py migrate
# At this point, the reversion tables are just like the ones at 1.10.2 django migrations
$ pip install "Django>=1.7" django-reversion==1.10.2
# Do the necessary changes to bump the django version
$ python manage.py migrate --fake-initial
$ pip install django-reversion==2.0.6
# This will apply the remaining migrations
$ python manage.py migrate
```

**Comment by Rikuoja** [comment]
+1 here, our app has the same problem with postgres. Will try out the above solution.

**Comment by etianen** [comment]
Is there anything that needs to be fixed with django-reversion here? From what I can tell, this is just a difficulty with the old south-style migrations migrating to the new builtin migrations.

**Comment by Rikuoja** [comment]
Hm, at least our app never used south, it was implemented with django only. Therefore, the problem is not specific to south.

**Comment by etianen** [comment]
Surely then this is a bug in the django migration framework. There's nothing going on in the django-reversion migations that is an abuse of the migration framework.

**Comment by Rikuoja** [comment]
Looks like removing `manager_slug` in migration 0003 was, for some reason, not executed on our postgres, when the migration was run together with migrations 0004 and 0001_squashed_0004. This happened when upgrading to django-reversion 2. Any idea why?

**Comment by etianen** [comment]
If that's the cause, then I'd suggest a bug in the migration framework.

This is tricky to fix in django-reversion, as all the migrations have been
in the wild for a considerable time. Worse, they work for most people,
making the bug really hard to reproduce.

**Comment by tclancy** [comment]
Adding another data point here: we have a Django project that was upgraded from 1.5.1 to 1.8.17 on MySQL and are running into this issue. We upgraded from `django-reversion==1.7.1` to `2.0.8` directly as part of this process and it looks like none of the migrations from 0002-0004 are applied. The migrations table shows 0001-0004 being applied as a group and then the squashed migration being applied after every other app had finished their migrations. I've tried removing all of the entries from `django_migrations` and running them again with and without the squashed file in the `migrations` folder, but the initial migration created for the upgrade to Django 1.8 lists `0001_squashed_0004_auto_20160611_1202` as a dependency. If I remove that the process still does not work.

**Comment by tclancy** [comment]
There seems to be some kind of ordering/ dependency problem: when I try to make a new, empty migration file for the app in order to correct this via hack, the file is prefixed as `0002`.

**Comment by BallisticPain** [comment]
So I've ended up running into this same issue... I've upgraded an application from Django 1.2 to 1.8.18 as of writing this comment... thankfully I've "pulled" staging and prod data locally to test the migrations as I was certainly concerned about it. I'll document my specific steps and see if I can arrive at a working solution. I must --fake-initial as my database already exists, but the new migration table doesn't. I guess as a library this may be one time a squash may not make as much sense. Going from something so old to much newer has been a royal pain. I'll keep this ticket around so I can pop back in and provide hopefully a documented working solution for my structures. I can test and re-test since I'm pulling it all local first. :-D

Thanks for the wonderful library btw...

**Comment by BallisticPain** [comment]
Okay... so one thing I've noticed as I'm playing around and learning is... the squashed migration was made with Django 1.9.7, and while this may work for Django versions before 1.9, it isn't guaranteed. For anyone else moving forward... I'm still working on a "works every time" step-by-step.

**Comment by am0s** [comment]
I had the same issue and as far as I can see the squashed migration is the reason for the problems. Any installation that earlier used south for migrations will need to use --fake-initial to install the new migration files, if not Django migrations will complain that the table already exist.
However as the squashed file replaced the initial one the entire squashed file will be faked, this means that the database schema will no longer match models.py definition.

I think it was a mistake to squash all the migrations, the `0001` migration should have been left out as it provided the first compatibility point for those with existing databases.

However to get around this the trick is to skip the squashed migration for now. The steps I used was:

```
# 1. Optional step, reset migrations to before the first migration
$ python manage.py migrate --fake reversion zero

# 2. Remove 0001_squashed_0004_auto_20160611_1202.py* from reversion/migrations, locate it in the site-packages folder in your virtual-env or global installation.

# 3. Run the first migration with --fake-initial, this marks the migration as being applied if the tables already exist.
$ python manage.py migrate --fake-initial reversion 0001

# 4. Run the rest of the migrations
$ python manage.py migrate reversion

# 5. Reinstall django-reversion to get the squashed file back, for install previous version and then back up again, for instance I used:
$ pip install django-reversion==2.0.12
$ pip install django-reversion==2.0.13
```

`showmigrations` should now show that the squashed file is applied (as it replaces all the existing migrations)

```
$ python manage.py showmigrations
reversion
 [X] 0001_squashed_0004_auto_20160611_1202 (4 squashed migrations)
```

**Comment by etianen** [comment]
Did reversion's squashed migrations in any way deviate from Django's supported and documented way of squashing revisions? If so, I'll be happy to fix it.

Otherwise, this is a bug in the Django migration framework, and needs to be investigated there.

**Comment by cericoda** [comment]
Not sure this will help anyone but I just ran into this upgrading from reversion 1.6.5/Django 1.4.

My solution ended up being:- 

- upgrade reversion to version 2.02, 
- move away/delete the squashed migration file 0001_squashed_0004_auto_20160611_1202 in that package, 
- fake the first migration
`python manage.py migrate --fake reversion 0001`
- migrate the other 3 migrations
`python manage.py migrate reversion`
- upgrade to version 3.0.0
- finally, fake the squashed migration
`python manage.py migrate --fake reversion`
